src/irmin-pack/inode.ml

```diff
@@ -870,6 +870,17 @@ struct
           seq_tree layout (Array.to_seq t.entries) ~depth ~cache k ~off ~len
       | Values vs -> seq_values layout (StepMap.to_seq vs) k ~off ~len
 
+    let list_v layout v ~cache k ~off ~len =
+      match v with
+      | Tree _ ->
+          let s () = seq_v layout v ~cache k ~off ~len in
+          List.of_seq s
+      | Values vs ->
+          if off = 0 && len = Int.max_int then StepMap.bindings vs
+          else
+            let seq () = seq_values layout (StepMap.to_seq vs) k ~off ~len in
+            List.of_seq seq
```

Comment:
Where is the allocation savings coming from if `List.of_seq` is still used for construction?

Reply:
The fast path is the line above

Reply:
Chatting with @Ngoguey42 last week shed light on this code for me. I'll record here for public record :smile: 

The `off` and `len` variables are used for pagination so line 879 is a fast path (it does not create intermediate `Seq.t` values) when no pagination is used.